### PR TITLE
Closes #3793: Improve accessibility of card link titles

### LINKED ISSUE #3793: Improve accessibility of card link titles
## Problem/Motivation
Issue found on a [Dubbot scan](https://app.dubbot.com/sites/5dc32d02482faf0915d31627/pages/6636d6ca337a96db153cbe7f?tab=accessibility&issueId=link-name).

### Describe the bug
When the Card Link URL field is filled out and the Card Link Title is _not_ populated, there is an accessibility concern because screen readers cannot tell what the link is.

### To Reproduce
1. Add a card to a page
2. Complete the Card Link URL field
3. Do not fill in the Card Link Title field
4. Save
5. Inspect the page and find the card
6. Observe no title or content around the `<a>` tag on the card

## Proposed resolution
1. Require the Card Link Title field -- Only when URL field is populated
2. Add a selection in the Card Link Style dropdown for `Hidden` or similar

### Expected behavior
I expect links to have accurate descriptive link titles.

## Additional context
Consider `sr-only` class. https://getbootstrap.com/docs/4.0/utilities/screenreaders/

Expect to create a database update to look for empty titles, with populated URLs. In this case, we should have a reasonable default. Maybe auto-populate the link title with the card title and select the hidden link style? 

Or just not allow save and have a notice with help text?

### PR DESCRIPTION
**Pull request by joeparsons**
## Description

- Adds a new "Hidden" card link tltle style option (`sr-only`)
- Adds validation to the card link title to require a value of a card link URL value is provided

### Release notes

```
This change makes the Card Link Title field required if the Card Link URL field is populated.
```

## Related issues

Closes #3793 

## How to test

Try adding cards to flexible pages using different title style options with and without card link title and card link URL values.

1. Add cards to a page
2. Test clickable and non-clickable card style with the following options
    a. With/out Card Link Title field
    b. With/out Card Link URL field
3. Repeat with title on image vs title within card styles

### Arizona Quickstart (install profile, custom modules, custom theme)
- **Patch release changes**
   - [ ] Bug fix
   - [x] Accessibility, performance, or security improvement
   - [ ] Critical institutional link or brand change
   - [ ] Adding experimental module
   - [ ] Update experimental module
- **Minor release changes**
   - [ ] New feature
   - [x] Breaking or visual change to existing behavior
   - [ ] Upgrade experimental module to stable
   - [ ] Enable existing module by default or database update
   - [ ] Non-critical brand change
   - [ ] New internal API or API improvement with backwards compatibility
   - [ ] Risky or disruptive cleanup to comply with coding standards
   - [ ] High-risk or disruptive change (requires upgrade path, risks regression, etc.)
- **Other or unknown**
   - [ ] Other or unknown

### Drupal core
- **Patch release changes**
   - [ ] Security update
   - [ ] Patch level release (non-security bug-fix release)
   - [ ] Patch removal that's no longer necessary
- **Minor release changes**
   - [ ] Major or minor level update
- **Other or unknown**
   - [ ] Other or unknown

### Drupal contrib projects
- **Patch release changes**
   - [ ] Security update
   - [ ] Patch or minor level update
   - [ ] Add new module
   - [ ] Patch removal that's no longer necessary
- **Minor release changes**
   - [ ] Major level update
- **Other or unknown**
   - [ ] Other or unknown

## Checklist

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.
- [ ] My change requires release notes.

### CODE REVIEW

File: modules/custom/az_demo/data/az_demo_card_paragraph.json [hunk 1]
```diff
@@ -51,13 +51,13 @@
           },
           {
             "delta": 1,
-            "title": "No link title",
+            "title": "Hidden link title",
             "body":"<p>Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua.</p>",
-            "link_title": "",
+            "link_title": "Hidden Card Link Title",
```
Comment (joeparsons) [inline modules/custom/az_demo/data/az_demo_card_paragraph.json:56]:
```suggestion
            "link_title": "Visually hidden card link",
```

File: modules/custom/az_demo/data/az_demo_card_paragraph.json [hunk 2]
```diff
@@ -92,13 +92,13 @@
         },
         {
           "delta": 1,
-          "title": "No link title",
+          "title": "Hidden link title",
           "body":"<p>Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua.</p>",
-          "link_title": "",
+          "link_title": "Hidden Card Link Title",
```
Comment (joeparsons) [inline modules/custom/az_demo/data/az_demo_card_paragraph.json:97]:
```suggestion
          "link_title": "Visually hidden card link",
```

**Comment by danahertzberg** [comment]
When the following selections are made, the link does not span the whole card.

* Clickable card
* Card link title and URL fields populated
* Hidden card link title

Example: https://7b5e2985-cedf-4a13-ab91-3a35916c835d--pr-3814.probo.build/dana

File: modules/custom/az_card/src/Plugin/Field/FieldWidget/AZCardWidget.php [hunk 3]
```diff
@@ -420,6 +433,20 @@ public function cardAjax(array &$form, FormStateInterface $form_state) {
...
+   * Form element validation handler for the 'link_title' field.
+   *
+   * Makes field required if link_uri is provided.
+   */
+  public function validateCardLinkTitle(&$element, FormStateInterface $form_state, &$complete_form) {
+    $parents = $element['#array_parents'];
+    array_pop($parents);
+    $parent_element = NestedArray::getValue($complete_form, $parents);
+    if (empty($element['#value']) && !empty($parent_element['link_uri']['#value'])) {
+      $form_state->setError($element, t('Card link title is required when a URL is provided.'));
```
Comment (danahertzberg) [inline modules/custom/az_card/src/Plugin/Field/FieldWidget/AZCardWidget.php:446]:
```suggestion
      $form_state->setError($element, t('Card Link Title field is required when a URL is provided. Card Link Title may be visually hidden with a Card Link Style selection.'));
```

File: modules/custom/az_card/src/Plugin/Field/FieldWidget/AZCardWidget.php [hunk 4]
```diff
@@ -261,8 +261,9 @@ public function formElement(FieldItemListInterface $items, $delta, array $elemen
     $element['link_title'] = [
       '#type' => 'textfield',
       '#title' => $this->t('Card Link Title'),
+      '#element_validate' => [[$this, 'validateCardLinkTitle']],
       '#default_value' => $item->link_title ?? NULL,
-      '#description' => $this->t('Make each link title unique for <a href="https://www.w3.org/WAI/WCAG21/Understanding/link-purpose-in-context.html">best accessibility</a> of this content. Use the pattern <em>"verb" "noun"</em> to create helpful links. For example, "Explore Undergraduate Programs".'),
+      '#description' => $this->t('<p>Make each link title unique for <a href="https://www.w3.org/WAI/WCAG21/Understanding/link-purpose-in-context.html">best accessibility</a> of this content. Use the pattern <em>"verb" "noun"</em> to create helpful links. For example, "Explore Undergraduate Programs".</p><p>This field is required when a Card Link URL is provided. Card Link Title may be visually hidden with Card Link Style selection.</p>'),
```
Comment (danahertzberg) [inline modules/custom/az_card/src/Plugin/Field/FieldWidget/AZCardWidget.php:266]:
```suggestion
      '#description' => $this->t('<p>Make each link title unique for <a href="https://www.w3.org/WAI/WCAG21/Understanding/link-purpose-in-context.html">best accessibility</a> of this content. Use the pattern <em>"verb" "noun"</em> to create helpful links. For example, "Explore Undergraduate Programs".</p><p>This field is required when a Card Link URL is provided. Card Link Title may be visually hidden with a Card Link Style selection.</p>'),
```

**Review by danahertzberg — APPROVED** [review]
Will make a follow up ticket for the `#` link going to the homepage instead of the current page

- https://github.com/az-digital/az_quickstart/issues/3821

**Review by ewlyman — APPROVED** [review]
This works as described.

**Comment by danahertzberg** [comment]
Current needs 
- Solid communication around updating content on sites

Will look into DB update for future
- If no link title, make card title and select hidden link title
- If no card title, TBD
   - Get title of destination link? Probs not

Future
- Look into providing accessibility report on the site admin side
- How do we communicate and fix issues easily across whole site?

**Pull request merged**